### PR TITLE
[libc++] Optimize std::find if types are integral and have the same signedness

### LINKED ISSUE #70238: libc++, find(char*, char*, int) does not optimize to memchr
Seems like this case was missed in recent changes. For int constants it seems important~ish, like  find 0.
https://godbolt.org/z/3r68MxWrc

Issue was found by @pdimov

### PR DESCRIPTION
**Pull request by philnik777**
Fixes #70238

### CODE REVIEW

**Review by ldionne — COMMENTED** [review]
Overall I think this makes sense, but it's tricky so we really need some tests.

Also, this is the kind of stuff that I would really expect something like Clang IR to render completely obsolete (along with most of the `memchr`-related optimizations, I guess). But we're not there yet I guess.

CC @bcardosolopes for awareness -- we have a bunch of these optimizations where we try to lower things like `std::find(char*, char*, char)` to a call to `memchr`, and we're thinking that Clang IR might make it easy for the compiler to understand that this optimization is possible without us telling it. In case you have thoughts about this, we'd love to hear them.

File: libcxx/test/std/algorithms/alg.nonmodifying/alg.find/find.pass.cpp [hunk 1]
```diff
@@ -113,6 +113,272 @@ struct TestTypes {
...
 };
 
+enum class match {
+  none,
+  first,
+};
+
+template <class T, class U>
+TEST_CONSTEXPR_CXX20 void test_integer_promotion(match min_expected, match max_expected) {
+  { // check min value
```
Comment (ldionne) [inline libcxx/test/std/algorithms/alg.nonmodifying/alg.find/find.pass.cpp:123]: Instead you could say:

```
bool should_match = T(value) == U(value);
```

and then it doesn't really matter what size different types are, you'd be checking that we have the same behavior that a "naive" implementation of `std::find` (using `==`) would have, which is really what we care about here.

**Review by ldionne — COMMENTED** [review]
IMO it would be reasonable to land this improvement without handling different signedness for now.

File: libcxx/test/std/algorithms/alg.nonmodifying/alg.find/find.pass.cpp (hunk 1 above)
Comment (ldionne) [reply libcxx/test/std/algorithms/alg.nonmodifying/alg.find/find.pass.cpp:123]: And then you can throw a bunch of different values at the test much more easily, I think.

**Comment by EricWF** [comment]
Are there benchmark results for this?

**Comment by philnik777** [comment]
> Are there benchmark results for this?

https://reviews.llvm.org/D144394#4138976

**Review by EricWF — APPROVED** [review]

File: libcxx/test/std/algorithms/alg.nonmodifying/alg.find/find.pass.cpp [hunk 2]
```diff
@@ -162,6 +163,39 @@ void test_deque() {
...
+struct TestIntegerPromotions1 {
+  template <class U>
+  TEST_CONSTEXPR_CXX20 void test(T val, U to_find) {
+    bool expect_match = val == to_find;
+    assert(std::find(&val, &val + 1, to_find) == (expect_match ? &val : &val + 1));
+  }
+
+  template <class U>
+  TEST_CONSTEXPR_CXX20 void operator()() {
+    test<U>(0, 0);
```
Comment (ldionne) [inline libcxx/test/std/algorithms/alg.nonmodifying/alg.find/find.pass.cpp:176]: We should test with other values than `0`.

**Review by ldionne — COMMENTED** [review]
I think this looks good to me if we have tests with other than `0`.

**Review by ldionne — APPROVED** [review]
Actually I am going to approve this but please add the non-zero tests, and ping me again if you need to change the implementation when you add these tests.

**Pull request merged**